### ISSUE TITLE
NPM for JXcore versioning

### ISSUE THREAD

**Issue by ktrzeciaknubisa** [jxcore/jxcore#475]
We keep NPM for JXcore under single folder `~/.jx/npm` and it is the same path per each jxcore version. That leads to a situation when older JXcore may overwrite it with its own downloaded package. I myself had few such cases when `jx install` didn't work as it should because of that reason.

1. I suggest to start keeping NPM in separate directories, like:

    ```bash
~/.jx/npm304
~/.jx/npm305
# etc.
```

2. It will also eliminate the need for manual `~/.jx` removal when sometimes we update NPM.
3. The `~/.jx/v 0.3.0.5` text files will no longer be needed.

How about this?

**Comment by plika** [comment]
This would end up keeping lots of folders on users' system

**Comment by ktrzeciaknubisa** [comment]
Agree. However something probably should be done anyway. At least to make sure, that each `jx install` regardless of jx version calls its own npm. Especially: newer jx should not call older npm.

If we don't want to multiplicate folders, than let's keep just one, but instead of writing `~/.jx/v 0.3.0.5` text files, let's mark which jx version downloaded the npm as last.

**Comment by jasonkarns** [comment]
Perhaps jxcore could follow node's lead w.r.t. how npm is bundled. npm is bundled under the `libexec` directory.

When installed from source, jxcore's installation directory uses the standard layout (bin/, lib/, man/). I propose that npm be dropped into `libexec/`. (Indeed, this is precisely how npm is bundled for homebrew)

A symlink can then be created under `bin/` (next to `jx`) that points to `libexec/npm/bin/npm-cli.js`
I propose that npm be installed under `libexec` in jxcore's installation directory.

Bonus: this would make npm work out of the box for jxcore versions managed with nodenv.

Super bonus: `~/.jx` wouldn't need to exist at all.

**Comment by ktrzeciaknubisa** [comment]
That's not just `~/.jx` . We also make a use of `~/.node-gyp` and put there our e.g. `jxb304` folder (used by our npm). But the bundling idea seems to be correct.

**Comment by jasonkarns** [comment]
Is the npm that's dropped under `~/.jx` modified in any way? Or is it just regular old npm? (Or maybe merely a specific version of npm?)

**Comment by ktrzeciaknubisa** [comment]
Yes, it is modified in many ways. The sources are under `deps/npm`.

**Comment by jasonkarns** [comment]
I see that the biggest difference is the shebang replaces node with jx. This won't be necessary from within nodenv since node will be symlinked to jx. What's the likelihood that regular old npm will work on jx within a nodenv environment where `node` is really `jx`? (no shebang alterations needed)

**Comment by ktrzeciaknubisa** [comment]
It's not just shebang. Some packages under node_modules were also altered, e.g. `node-gyp`. Also we force ` --build-from-source` (for node-pre-gyp), if a module installs downloadable prebuilt binaries compiled against node. Couple of other things too, e.g. `--autoremove` option useful for mobile development.

BTW. replacing `jx` into `node` may case failures when running apps with native addons built by `jx install`.

**Comment by jasonkarns** [comment]
> replacing jx into node may case failures

Keep in mind that in the context of nodenv, `node` is just a symlink to `jx`
